python/taichi/lang/impl.py

```diff
@@ -300,7 +300,44 @@ def layout(func):
 
 
 def ti_print(*vars):
-    taichi_lang_core.create_print([Expr(var).ptr for var in vars])
+    def ent2content(var):
+        if isinstance(var, str):
+            return var
+        else:
+            return Expr(var).ptr
+
+    def vars2ents(vars):
+        for var in vars:
+            if hasattr(var, '__ti_repr__'):
+                repr = var.__ti_repr__()
+                for v in vars2ents(repr):
+                    yield v
+            else:
+                yield var
+
+    def add_separators(vars):
+        for var in vars:
+            yield var
+            yield ' '
+
+    def fuse_str(ents):
+        stack = ''
+        for ent in ents:
+            if isinstance(ent, str):
+                stack += ent
+            else:
+                if stack:
+                    yield stack
+                    stack = ''
+                yield ent
+        if stack:
+            yield stack
```

Comment:
This is too much confusion. `stack` here doesn't really look like a stack since you push but never pop. A name like `fused_string` is probably better.